tools/firmware-utils/src/mkzyxelzldfw.c

```diff
@@ -0,0 +1,835 @@
+	header_p->files_offset = ntohs(header_p->files_offset);
+	header_p->header_length = ntohs(header_p->header_length);
+	header_p->info_length = ntohs(header_p->info_length);
+}
+
+static void translate_file_header(struct fw_header_file *file_header)
+{
+	file_header->type = ntohs(file_header->type);
+	file_header->flags = ntohs(file_header->flags);
+	file_header->length = ntohl(file_header->length);
+	file_header->checksum = ntohl(file_header->checksum);
+	file_header->flash_offset = ntohl(file_header->flash_offset);
+}
+
+static void translate_kernel_header(struct fw_header_kernel *kernel_header)
+{
+	int i;
+
+	kernel_header->bm_checksum = ntohl(kernel_header->bm_checksum);
+	kernel_header->kernel_checksum = ntohl(kernel_header->kernel_checksum);
+	/*kernel_header->zld_checksum = ntohl(kernel_header->zld_checksum);*/
+	kernel_header->core_checksum = ntohl(kernel_header->core_checksum);
+
+	kernel_header->max_models = ntohs(kernel_header->max_models);
+	for (i = 0; i < MAX_MODELS; i++)
+		kernel_header->models[i] = ntohs(kernel_header->models[i]);
+	kernel_header->baudrate = ntohl(kernel_header->baudrate);
+}
+
+static void checksum_add_from_buf(MD5_CTX *ctx, void *buf,
+			    size_t length, size_t offset)
+{
+	char *begin = &((char *)buf)[offset];
+
+	MD5_Update(ctx, begin, length);
+}
+
+static void checksum_add_from_file(MD5_CTX *ctx, FILE *fp_src,
+			    size_t length, size_t offset)
+{
+	int buf[512];
+	size_t len;
+
+	fseek(fp_src, (long int)offset, SEEK_SET);
+	while (length) {
+		if (length >= sizeof(buf)) {
+			len = sizeof(buf);
+		} else {
+			len = length;
+		}
+
+		length -= len;
+
+		if (fread(buf, len, 1, fp_src) != 1)
+			error("Failed to read for checksum calculation");
+
+		checksum_add_from_buf(ctx, buf, len, 0);
+	}
+}
+
+static uint32_t checksum_finish(MD5_CTX *ctx)
+{
+	unsigned char md5sum[16];
+	uint32_t checksum = 0;
+	int i;
+
+	MD5_Final(md5sum, ctx);
+
+	for (i = 0; i < 16; i += 4)
+		checksum += ((uint32_t)md5sum[i] << 24 |
+			     (uint32_t)md5sum[i + 1] << 16 |
+			     (uint32_t)md5sum[i + 2] << 8 |
+			     (uint32_t)md5sum[i + 3]);
+
+	if (checksum == 0)
+		checksum = 1;
+
+	return checksum;
+}
+
+static uint32_t checksum_calculate(FILE *fp, size_t kernel_offset)
+{
+	struct fw_header_kernel dummy;
+	MD5_CTX ctx;
+	size_t file_size;
+
+	fseek(fp, 0, SEEK_SET);
+	file_size = get_file_size(fp);
+
+	MD5_Init(&ctx);
+
+	checksum_add_from_file(&ctx, fp, kernel_offset - CHECKSUM_SIZE,
+			       CHECKSUM_SIZE);
+
+	/* use a zeroed out kernel version header */
+	memset(&dummy, 0, sizeof(dummy));
+	checksum_add_from_buf(&ctx, &dummy, sizeof(dummy), 0);
+
+	checksum_add_from_file(&ctx, fp,
+			       file_size - kernel_offset - sizeof(dummy),
+			       kernel_offset + sizeof(dummy));
+
+	return checksum_finish(&ctx);
+}
+
+static uint32_t checksum_calculate_file(char *filename)
+{
+	MD5_CTX ctx;
+	FILE *fp;
+	size_t file_size;
+
+	if (access(filename, R_OK) || !(fp = fopen(filename, "rb")))
```

Comment:
Is the `access` check needed here? If it fails `fopen` should also fail.

Reply:
Removed the check.